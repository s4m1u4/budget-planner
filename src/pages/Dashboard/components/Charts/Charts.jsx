import React, { useState } from "react";
import { ChartsPie } from "./ChartsPie";
import { ChartsForm } from "./ChartsForm";
import { IconButton } from "@mui/material";
import {
  calculateDataIncome,
  calculateDataExpense,
  calculateRecordsIncome,
  calculateRecordsExpense,
  calculateCategoriesIncome,
  calculateCategoriesExpense,
} from "./helpers";
import AddCircleOutlineSharpIcon from "@mui/icons-material/AddCircleOutlineSharp";

import {
  Section,
  SectionBody,
  SectionHeader,
  SectionTitle,
} from "../../Dashboard.styles";
import { ChartsList } from "./Charts.styles";

export const Charts = ({ setNewCategory, records }) => {
  const [isOpenModal, setIsOpenModal] = useState(false);
  const handleOpen = () => setIsOpenModal(true);
  const handleClose = () => setIsOpenModal(false);

  const recordIncome = calculateRecordsIncome(records);
  const recordsExpense = calculateRecordsExpense(records);

  const categoriesIncome = calculateCategoriesIncome(recordIncome);
  const categoriesExpense = calculateCategoriesExpense(recordsExpense);

  const dataIncome = calculateDataIncome(categoriesIncome, recordIncome);
  const dataExpense = calculateDataExpense(categoriesExpense, recordsExpense);

  return (
    <Section area="charts">
      <SectionHeader>
        <SectionTitle>Charts</SectionTitle>
        <IconButton size="small" sx={{ padding: "0" }} onClick={handleOpen}>
          <AddCircleOutlineSharpIcon />
        </IconButton>
      </SectionHeader>
      <SectionBody>
        <ChartsList>
          <ChartsPie title="Income" data={dataIncome} />
          <ChartsPie title="Expense" data={dataExpense} />
        </ChartsList>
      </SectionBody>
      <ChartsForm
        open={isOpenModal}
        handleClose={handleClose}
        setNewCategory={setNewCategory}
      />
    </Section>
  );
};
