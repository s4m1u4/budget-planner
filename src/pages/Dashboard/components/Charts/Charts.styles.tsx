import styled from "@emotion/styled";

export const ChartsList = styled.div({
  display: "grid",
  gridTemplateColumns: "1fr 1fr",
});
