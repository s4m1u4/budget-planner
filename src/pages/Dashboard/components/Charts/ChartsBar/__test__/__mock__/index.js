export { TITLE, DATA } from "./mock";
