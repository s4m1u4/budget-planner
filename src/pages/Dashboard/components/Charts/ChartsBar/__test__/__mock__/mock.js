export const TITLE = "Income";
export const DATA = [
  { name: "Work", income: 17369, expense: -13223 },
  { name: "Travel", income: 74123, expense: -84534 },
  { name: "Education", income: 49763, expense: -2533 },
];
