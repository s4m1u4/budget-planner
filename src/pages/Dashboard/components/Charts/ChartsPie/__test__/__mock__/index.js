export { TITLE, DATA_INCOME } from "./mock";
