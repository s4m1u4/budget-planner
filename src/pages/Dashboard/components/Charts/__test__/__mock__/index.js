export { CATEGORIES, RECORDS } from "./mock";
