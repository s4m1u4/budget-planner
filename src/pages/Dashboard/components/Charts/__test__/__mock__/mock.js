export const RECORDS = [
  {
    id: "9cc6c0d7-24e3-46aa-aa7c-73e86c1dd230",
    amount: "2354.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
  {
    id: "bc732adb-7b27-4d76-944c-4e7013eb97fa",
    amount: "836.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
  {
    id: "9410db09-8e1c-4c76-a6b3-e766d9b53ae5",
    amount: "2454.0000",
    type: "income",
    user_id: "fe719e6f-ac63-4630-9a70-2112037f9078",
    category_id: "084a3558-6b14-4e38-9d24-47e912021384",
  },
];

export const CATEGORIES = [
  {
    id: "084a3558-6b14-4e38-9d24-47e912021384",
    title: "Work",
    description: "Work",
    isDefault: false,
  },
  {
    id: "0ce314b9-a547-4f57-9d64-e466777ba8e9",
    title: "Travel",
    description: "awdawda",
    isDefault: false,
  },
  {
    id: "ea298d16-8bdb-4c5b-8198-6b2e5ce2f9a2",
    title: "Education",
    description: "Education it's so important for you!",
    isDefault: false,
  },
];
