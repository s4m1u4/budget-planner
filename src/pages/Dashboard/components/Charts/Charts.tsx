import React, { FC, useEffect, useState } from "react";
import { ChartsPie } from "./ChartsPie";
import { ChartsBar } from "./ChartsBar";
import { ChartsForm } from "./ChartsForm";
import { IconButton } from "@mui/material";
import {
  calculateDataIncome,
  calculateDataExpense,
  calculateRecordsIncome,
  calculateRecordsExpense,
  calculateCategoriesIncome,
  calculateCategoriesExpense,
  calculateUniqueCategories,
  calculateDataChartsBar,
} from "./helpers";
import PieChartIcon from "@mui/icons-material/PieChart";
import BarChartIcon from "@mui/icons-material/BarChart";
import AddCircleOutlineSharpIcon from "@mui/icons-material/AddCircleOutlineSharp";
import DeleteForeverRoundedIcon from "@mui/icons-material/DeleteForeverRounded";
import { ICategoryData } from "../../types";
import { ICategory, IRecord } from "../../../../types";
import { ChartsModal } from "./ChartsModal";

import {
  Section,
  SectionBody,
  SectionHeader,
  SectionTitle,
} from "../../Dashboard.styles";
import { ChartsList } from "./Charts.styles";
import { IconsGroup } from "../../../Records/Records.styles";

interface ChartsProps {
  records: IRecord[];
  categories: ICategory[];
  deleteCategory: (id: string) => void;
  setNewCategory: (categoryData: ICategoryData) => void;
}

export const Charts: FC<ChartsProps> = ({
  setNewCategory,
  deleteCategory,
  records,
  categories,
}) => {
  useEffect(() => {
    setIsPieChart(
      sessionStorage.getItem("charts") === "pie" ||
        sessionStorage.getItem("charts") === null
        ? true
        : false
    );
  }, []);

  const [isPieChart, setIsPieChart] = useState<boolean>(true);
  const [isOpenModalCreate, setIsOpenModalCreate] = useState<boolean>(false);
  const [isOpenModalDelete, setIsOpenModalDelete] = useState<boolean>(false);
  const handleOpenCreate = () => setIsOpenModalCreate(true);
  const handleCloseCreate = () => setIsOpenModalCreate(false);
  const handleOpenDelete = () => setIsOpenModalDelete(true);
  const handleCloseDelete = () => setIsOpenModalDelete(false);

  const recordIncome: IRecord[] = calculateRecordsIncome(records);
  const recordsExpense: IRecord[] = calculateRecordsExpense(records);

  const categoriesIncome = calculateCategoriesIncome(recordIncome);
  const categoriesExpense = calculateCategoriesExpense(recordsExpense);

  const dataIncome = calculateDataIncome(categoriesIncome, recordIncome);
  const dataExpense = calculateDataExpense(categoriesExpense, recordsExpense);

  const uniqueCategories = calculateUniqueCategories(
    categoriesIncome,
    categoriesExpense
  );
  const dataChartsBar = calculateDataChartsBar(
    uniqueCategories,
    dataIncome,
    dataExpense
  );

  const handleClickPieCharts = () => {
    sessionStorage.setItem("charts", "pie");
    setIsPieChart(true);
  };

  const handleClickBarCharts = () => {
    sessionStorage.setItem("charts", "bar");
    setIsPieChart(false);
  };

  return (
    <Section area="charts">
      <SectionHeader>
        <SectionTitle>Charts</SectionTitle>
        <IconsGroup>
          <IconButton
            size="small"
            sx={{ padding: "0" }}
            data-charts="charts-pie"
            onClick={handleClickPieCharts}
          >
            <PieChartIcon />
          </IconButton>
          <IconButton
            size="small"
            sx={{ padding: "0" }}
            data-charts="charts-bar"
            onClick={handleClickBarCharts}
          >
            <BarChartIcon />
          </IconButton>
          <IconButton
            size="small"
            sx={{ padding: "0" }}
            data-modal="modal-create"
            onClick={handleOpenCreate}
          >
            <AddCircleOutlineSharpIcon />
          </IconButton>
          <IconButton
            size="small"
            sx={{ padding: 0 }}
            data-modal="modal-delete"
            onClick={handleOpenDelete}
            disabled={!Boolean(categories.length)}
          >
            <DeleteForeverRoundedIcon />
          </IconButton>
        </IconsGroup>
      </SectionHeader>
      <SectionBody>
        {isPieChart ? (
          <ChartsList>
            <ChartsPie title="Income" data={dataIncome} />
            <ChartsPie title="Expense" data={dataExpense} />
          </ChartsList>
        ) : (
          <ChartsBar data={dataChartsBar} />
        )}
      </SectionBody>
      <ChartsForm
        open={isOpenModalCreate}
        onSubmit={(values) => values}
        handleClose={handleCloseCreate}
        setNewCategory={setNewCategory}
      />
      <ChartsModal
        open={isOpenModalDelete}
        onSubmit={(values) => values}
        handleClose={handleCloseDelete}
        categories={categories}
        deleteCategory={deleteCategory}
      />
    </Section>
  );
};
