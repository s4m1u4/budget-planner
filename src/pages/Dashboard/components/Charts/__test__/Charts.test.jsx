import { Charts } from "../Charts";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { CATEGORIES, RECORDS } from "./__mock__";

const renderComponent = () => {
  return render(
    <Charts
      records={RECORDS}
      categories={CATEGORIES}
      deleteCategory={deleteCategory}
      setNewCategory={setNewCategory}
    />
  );
};

const deleteCategory = jest.fn();
const setNewCategory = jest.fn();

describe("Charts component", () => {
  it("snapshot matching", () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  describe("should render", () => {
    beforeEach(() => {
      renderComponent();
    });

    it("the title", () => {
      expect(screen.getByText(/charts/i)).toBeInTheDocument();
    });
  });

  describe('modal "Create a new category"', () => {
    it("should open", () => {
      const { container } = renderComponent();
      const buttonOpenModal = container.querySelector(
        '[data-modal="modal-create"]'
      );

      userEvent.click(buttonOpenModal);

      expect(
        screen.getByRole("heading", { name: /create a new category/i })
      ).toBeInTheDocument();
    });

    it("should close", () => {
      const { container } = renderComponent();
      const buttonOpenModal = container.querySelector(
        '[data-modal="modal-create"]'
      );

      userEvent.click(buttonOpenModal);

      expect(
        screen.getByRole("heading", { name: /create a new category/i })
      ).toBeInTheDocument();

      userEvent.click(screen.getByRole("button", { name: /close/i }));

      expect(
        screen.queryByRole("heading", { name: /create a new category/i })
      ).not.toBeInTheDocument();
    });
  });

  describe('modal "Delete category"', () => {
    it("should open", () => {
      const { container } = renderComponent();
      const buttonOpenModal = container.querySelector(
        '[data-modal="modal-delete"]'
      );

      userEvent.click(buttonOpenModal);

      expect(
        screen.getByRole("heading", {
          name: /delete category/i,
        })
      ).toBeInTheDocument();
    });

    it("should close", () => {
      const { container } = renderComponent();
      const buttonOpenModal = container.querySelector(
        '[data-modal="modal-delete"]'
      );

      userEvent.click(buttonOpenModal);

      expect(
        screen.getByRole("heading", {
          name: /delete category/i,
        })
      ).toBeInTheDocument();

      userEvent.click(screen.getByRole("button", { name: /close/i }));

      expect(
        screen.queryByRole("heading", { name: /delete category/i })
      ).not.toBeInTheDocument();
    });
  });

  describe("render charts", () => {
    it("on click button 'Charts pie'", () => {
      const { container } = renderComponent();
      const buttonPieCharts = container.querySelector(
        '[data-charts="charts-pie"]'
      );

      userEvent.click(buttonPieCharts);

      expect(screen.getByText(/income/i)).toBeInTheDocument();
    });

    it("on click button 'Charts bar'", () => {
      const { container } = renderComponent();
      const buttonPieCharts = container.querySelector(
        '[data-charts="charts-bar"]'
      );

      userEvent.click(buttonPieCharts);

      expect(screen.queryByText(/income/i)).not.toBeInTheDocument();
    });
  });
});
