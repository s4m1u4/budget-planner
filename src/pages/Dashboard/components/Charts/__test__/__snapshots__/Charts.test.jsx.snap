// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Charts component snapshot matching 1`] = `
<body>
  <div>
    <div
      class="css-a5nrzb"
    >
      <div
        class="css-13o2qpx"
      >
        <h3
          class="css-1xqtn9i"
        >
          Charts
        </h3>
        <div
          class="css-i5q2k0"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
            data-charts="charts-pie"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="PieChartIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M11 2v20c-5.07-.5-9-4.79-9-10s3.93-9.5 9-10zm2.03 0v8.99H22c-.47-4.74-4.24-8.52-8.97-8.99zm0 11.01V22c4.74-.47 8.5-4.25 8.97-8.99h-8.97z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
            data-charts="charts-bar"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="BarChartIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 9.2h3V19H5zM10.6 5h2.8v14h-2.8zm5.6 8H19v6h-2.8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1mgsdk9-MuiButtonBase-root-MuiIconButton-root"
            data-modal="modal-create"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AddCircleOutlineSharpIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-npdkjk-MuiButtonBase-root-MuiIconButton-root"
            data-modal="modal-delete"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DeleteForeverRoundedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm3.17-6.41a.9959.9959 0 0 1 0-1.41c.39-.39 1.02-.39 1.41 0L12 12.59l1.41-1.41c.39-.39 1.02-.39 1.41 0s.39 1.02 0 1.41L13.41 14l1.41 1.41c.39.39.39 1.02 0 1.41s-1.02.39-1.41 0L12 15.41l-1.41 1.41c-.39.39-1.02.39-1.41 0a.9959.9959 0 0 1 0-1.41L10.59 14l-1.42-1.41zM18 4h-2.5l-.71-.71c-.18-.18-.44-.29-.7-.29H9.91c-.26 0-.52.11-.7.29L8.5 4H6c-.55 0-1 .45-1 1s.45 1 1 1h12c.55 0 1-.45 1-1s-.45-1-1-1z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="css-1gk4ta0"
      >
        <div
          class="recharts-responsive-container"
          style="width: 99%; height: 330px;"
        />
      </div>
    </div>
  </div>
</body>
`;
